File: app/src/main/java/com/foxxyz/chronovision/InputStreamRequestBody.java
package com.foxxyz.chronovision;

import android.content.ContentResolver;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.BufferedSink;
import okio.Okio;

public class InputStreamRequestBody extends RequestBody {
    private final MediaType contentType;
    private final ContentResolver contentResolver;
    private final Uri uri;

    public InputStreamRequestBody(MediaType contentType, ContentResolver contentResolver, Uri uri) {
        if (uri == null) throw new NullPointerException("Uri is null");
        this.contentType = contentType;
        this.contentResolver = contentResolver;
        this.uri = uri;
    }

    @Nullable
    @Override
    public MediaType contentType() {
        return contentType;
    }

    @Override
    public long contentLength() throws IOException {
        return -1;
    }

    @Override
    public void writeTo(@NonNull BufferedSink sink) throws IOException {
        sink.writeAll(Okio.source(contentResolver.openInputStream(uri)));
    }
}

File: app/src/main/java/com/foxxyz/chronovision/InputStreamRequestBody.kt
package com.foxxyz.chronovision

import android.content.ContentResolver
import android.net.Uri

import java.io.IOException

import okhttp3.MediaType
import okhttp3.RequestBody
import okio.BufferedSink
import okio.source

class InputStreamRequestBody(private val contentType: MediaType, private val contentResolver: ContentResolver, private val uri: Uri?) : RequestBody() {

    init {
        if (uri == null) throw NullPointerException("Uri is null")
    }

    override fun contentType(): MediaType? {
        return contentType
    }

    @Throws(IOException::class)
    override fun contentLength(): Long {
        return -1
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        sink.writeAll(contentResolver.openInputStream(uri).source())
    }
}
